### PR TITLE
New unified StreamField & rich text block picker

### PR DESCRIPTION
**Pull request by thibaudcolas**
Implements two enhancements:

- #9131 – using the same combobox UI component to select rich text options and StreamField blocks in rich text, and StreamField without rich text
- #9754 – changing this combobox component to be more space-efficient

And fixes two bugs:

- "Position of add block triggers" item from #9755
- ~The blocks’ maximum number and current count not displaying in those block pickers~

Screen recording of the different variants of the picker:

block-picker

### ComboBox

This introduces a new ComboBox React component built with [Downshift](https://www.downshift-js.com/). This component is used in Draftail itself for the same UI. Now that we use this outside rich text in Wagtail, it makes much more sense to have the component defined here, and then configure the different widgets to use it.

- Downshift integration is 99% as-is from Draftail (only difference is better TypeScript types)
- Logic to filtering items is also 99% as-is, with the only difference also being better type annotations
- The appearance of the component is 80% retained. Differences are how we create the different `optgroup` within the menu from categories of items, and the two-column layout.

### Draftail integration

Not much to see there – there is a new prop where this component needs to be used, the component can be provided there so we use the same UI as with StreamField.

### StreamField integration

Replacing the previous block menu meant removing quite a bit of code, and altering the logic for what is allowed to be added when:

- We no longer restrict users from adding blocks of a _specific_ type when that has reached its maximum count
- We still restrict adding blocks when the whole field would go over its count

I believe we might want to remove that second restriction as well, but decided not to proceed as it wasn’t required to change this to proceed with the two specific changes here. 

---

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
-  ~~[ ] For Python changes: Have you added tests to cover the new/fixed behaviour?~~
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [x] **Please list the exact browser and operating system versions you tested**: Safari 16, Chrome 108, Firefox 108 on macOS 13.1. Safari iOS 16
    -   [x] **Please list which assistive technologies [^3] you tested**: VoiceOver macOS, WHCM, VoiceOver iOS
-   ~~[ ] For new features: Has the documentation been updated accordingly?~~

---

To test this, I’d recommend creating a new bakerydemo RecipePage to get access to all the different variations of the UI:

- The "Add" button should be visible when fields are empty
- In rich text, rich text options should be displayed without any grouping (Preface field on recipe page, or "Thank you text" field on form page in bakerydemo)
- In StreamField, StreamField blocks should appear either with no grouping, or the grouping defined by developers (Body field of RecipePage)
- In rich text within StreamField, rich text options appear first without a heading, then StreamField blocks either with their group or with a "StreamField blocks" group.

There is one bug remaining here which I chose to not tackle as part of this – for rich text inside StreamField, there should be a "Horizontal rule" option within the combobox, which doesn’t appear.

### CODE REVIEW

File: client/src/entrypoints/admin/telepath/widgets.js [hunk 1]
```diff
@@ -180,15 +180,15 @@ class DraftailInsertBlockCommand {
...
+    this.icon = blockDef.meta.icon;
     this.description = blockDef.meta.label;
     this.type = blockDef.name;
   }
 
   render({ option }) {
     // If the specific block has a limit, render the current number/max alongside the description
     const limitText =
-      typeof blockMax === 'number'
+      typeof this.blockMax === 'number'
```
Comment (thibaudcolas) [inline client/src/entrypoints/admin/telepath/widgets.js:191]: Unrelated bug fix – I’m surprised this isn’t being picked up by linting. Without `this.`, this code can’t work.

File: client/src/components/StreamField/blocks/StreamBlock.test.js [hunk 2]
```diff
@@ -517,7 +517,8 @@ describe('telepath: wagtail.blocks.StreamBlock with labels that need escaping',
 
   test('it renders correctly', () => {
     boundBlock.inserters[0].open();
-    expect(document.body.innerHTML).toMatchSnapshot();
+    const listbox = document.querySelector('[role="listbox"]');
+    expect(listbox.innerHTML).toMatchSnapshot();
```
Comment (thibaudcolas) [inline client/src/components/StreamField/blocks/StreamBlock.test.js:521]: I’ve changed this so the snapshot is a bit more specific about what we’re testing here (HTML char escaping)

File: client/src/components/StreamField/blocks/StreamBlock.js [hunk 3]
```diff
@@ -69,145 +77,99 @@ class StreamBlockMenu extends BaseInsertionControl {
...
-    });
-
-    this.outerContainer = dom.find('[data-streamblock-menu-outer]');
-    this.innerContainer = dom.find('[data-streamblock-menu-inner]');
-    this.hasRenderedMenu = false;
-    this.isOpen = false;
+    this.addButton = dom.find('button');
+    this.combobox = document.createElement('div');
     this.canAddBlock = true;
     this.disabledBlockTypes = new Set();
```
Comment (thibaudcolas) [inline client/src/components/StreamField/blocks/StreamBlock.js:93]: I’ve decided to retain this tracking of disabled block types as-is, even though the UI currently does nothing with i. This will be used again once we display the min/max block count (right now our current designs only show max).

File: client/src/components/Draftail/index.js [hunk 4]
```diff
@@ -205,8 +211,12 @@ const initEditor = (selector, originalOptions, currentScript) => {
         </>
       ),
       bottomToolbar: MetaToolbar,
-      commandPalette: (props) => (
-        <CommandPalette {...props} noResultsText={gettext('No results')} />
+      commandToolbar: (props) => (
```
Comment (thibaudcolas) [inline client/src/components/Draftail/index.js:214]: `commandPalette` was never the correct prop 😭 Until now the consequence was only the "no results text" being wrong and not translate-able.

File: client/src/components/ComboBox/findMatches.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+// Language-sensitive string comparison.
```
Comment (thibaudcolas) [inline client/src/components/ComboBox/findMatches.ts:1]: This whole file is a copy-paste of what we used in Draftail in v4.0 and v4.1, and the collator / contains setup are heavily taken from the different projects linked to below.

File: client/src/components/ComboBox/ComboBox.scss [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+.w-combobox__optgroup {
+  display: grid;
+  grid-template-columns: 1fr 1fr;
+  grid-auto-flow: column;
+  gap: theme('spacing.[0.5]');
+  padding: $spacing;
+  padding-top: 0;
+
+  @include media-breakpoint-up(sm) {
```
Comment (laymonage) [inline client/src/components/ComboBox/ComboBox.scss:47]: The spacing on top of the optgroup label seems a bit off on smaller screens, is this intended?

## Smaller screens

image

## Larger screens

image

**Review by laymonage — COMMENTED** [review]
Generally happy with this, it's so nice to finally have the combobox outside of the rich text field!

I only have one small comment on the spacing of the optgroup label, and I don't know how I can test this bugfix:

> The blocks’ maximum number and current count not displaying in those block pickers

**Review by laymonage — APPROVED** [review]
Managed to test the max block count, though as per discussion we'll leave out the fix for now as it requires more work to also show it on the block picker outside the rich text field. Apart from the spacing issue, this looks good to me. Thanks!

**Comment by thibaudcolas** [comment]
Thank you @laymonage 🤘 I’ve fixed this spacing issue, and took out that fix, can always add it back later if we agree it’s a good idea.

**Comment by th3hamm0r** [comment]
Awesome work!
Minor improvement: It would be nice, when similar to the richtext's block picker, the search field gets focused when the StreamField's block picker gets opened, so you can immediately start typing/searching.
Should I open an improvement-issue for that?
